### PR TITLE
KEYCLOAK-14973 Add GroupStorageManager

### PR DESCRIPTION
**Pull request by mhajas**

### CODE REVIEW

File: services/src/main/java/org/keycloak/storage/AbstractStorageManager.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        return getStorageProviderInstance(toStorageProviderModelType(componentModel));
+    }
+
+    /**
+     * Returns an instance of provider for the model
+     * @param model StorageProviderModel obtained from database/storage
+     * @return an instance of type CreatedProviderType
+     */
+    protected ProviderType getStorageProviderInstance(StorageProviderModelType model) {
+        // TODO: Add type checking.
```
[the remark this answers is not in the document]
Comment (mhajas) [reply services/src/main/java/org/keycloak/storage/AbstractStorageManager.java:135]: I agree I will remove the TODO comment.

**Comment by mhajas** [comment]
@hmlnarik, thank you for the review. I removed the TODO comment and squashed commits.

File: services/src/main/java/org/keycloak/storage/GroupStorageManager.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import org.keycloak.utils.ServicesUtils;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class GroupStorageManager extends AbstractStorageManager<GroupStorageProvider, GroupStorageProviderModel> implements GroupProvider {
+
+    public GroupStorageManager(KeycloakSession session) {
+        super(session);
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/storage/GroupStorageManager.java:37]: Could be 
```suggestion
        super(session, GroupStorageProvider.class, GroupStorageProviderModel::new);
```

File: services/src/main/java/org/keycloak/storage/AbstractStorageManager.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    /**
+     * Timeouts are used as time boundary for obtaining models from an external storage. Default value is set
+     * to 3000 milliseconds and it's configurable.
+     */
+    private static final Long STORAGE_PROVIDER_DEFAULT_TIMEOUT = 3000L;
+    protected final KeycloakSession session;
+    private Long storageProviderTimeout;
+
+    public AbstractStorageManager(KeycloakSession session) {
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/storage/AbstractStorageManager.java:49]: Could be 
```suggestion
    public AbstractStorageManager(KeycloakSession session, Class<ProviderType> providerTypeClass, Function<ComponentModel, StorageProviderModelType> componentToStorageProvider) {
```

**Review by hmlnarik — DISMISSED** [review]
@mhajas Looks good, I wondered about only one nit, WDYT?

**Comment by hmlnarik** [comment]
Thanks @mhajas! @vramik, could you please review as well?

File: services/src/main/java/org/keycloak/storage/GroupStorageManager.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * Obtaining groups from an external client storage is time-bounded. In case the external group storage
+     * isn't available at least groups from a local storage are returned. For this purpose
+     * the {@link org.keycloak.services.DefaultKeycloakSessionFactory#getClientStorageProviderTimeout()} property is used.
+     * Default value is 3000 milliseconds and it's configurable.
+     * See {@link org.keycloak.services.DefaultKeycloakSessionFactory} for details.
+     *
+     * TODO: Remove stream() and .collect(Collectors.toList()) during streamification
```
Comment (vramik) [inline services/src/main/java/org/keycloak/storage/GroupStorageManager.java:61]: Is this still valid?

**Review by vramik — COMMENTED** [review]
@mhajas thanks for the PR, it looks good to me, I have just one question.

File: services/src/main/java/org/keycloak/storage/GroupStorageManager.java (hunk 4 above)
Comment (mhajas) [reply services/src/main/java/org/keycloak/storage/GroupStorageManager.java:61]: I think so. This PR: https://github.com/keycloak/keycloak/pull/7328 should be merged first and then I will resolve this TODO

Comment (vramik) [reply services/src/main/java/org/keycloak/storage/GroupStorageManager.java:61]: oh I see, I thought this is another way around, thanks

**Review by vramik — DISMISSED** [review]

File: services/src/main/java/org/keycloak/storage/GroupStorageManager.java (hunk 4 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/storage/GroupStorageManager.java:61]: #7328 has been merged just now.

**Comment by mhajas** [comment]
Thank you for reviews @hmlnarik and @vramik. I rebased on the latest master, could you have another look, please?

**Review by hmlnarik — DISMISSED** [review]

File: server-spi-private/src/main/java/org/keycloak/storage/group/GroupStorageProviderFactory.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     * @param session
+     * @param realm
+     * @param model
+     */
+    @Override
+    default void onCreate(KeycloakSession session, RealmModel realm, ComponentModel model) {
+    }
+
+    /**
+     * configuration properties that are common across all RoleStorageProvider implementations
```
Comment (sguilhen) [inline server-spi-private/src/main/java/org/keycloak/storage/group/GroupStorageProviderFactory.java:94]: nitpicking: shouldn't the comment read "acrsoss all GroupStorageProvider implementations" instead?

File: testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/HardcodedGroupStorageProviderFactory.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    public static final String DELAYED_SEARCH = "delayed_search";
+
+    protected static final List<ProviderConfigProperty> CONFIG_PROPERTIES;
+
+    static {
+        CONFIG_PROPERTIES = ProviderConfigurationBuilder.create()
+                .property().name(GROUP_NAME)
+                .type(ProviderConfigProperty.STRING_TYPE)
+                .label("Hardcoded Group Name")
+                .helpText("Only this group naem is available for lookup")
```
Comment (sguilhen) [inline testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/HardcodedGroupStorageProviderFactory.java:45]: nitpicking: "name"

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/GroupStorageTest.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+            RealmModel realm = session.realms().getRealmByName("test");
+            StorageId storageId = new StorageId(providerId, "hardcoded-group");
+            GroupModel hardcoded = session.groups().getGroupById(realm, storageId.getId());
+            assertNotNull(hardcoded);
+        });
+    }
+
+    @Test(timeout = 4000)
+    public void testSearchTimeout() {
+        String hardcodedRole = HardcodedGroupStorageProviderFactory.PROVIDER_ID;
```
Comment (sguilhen) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/GroupStorageTest.java:94]: hardcodedGroup?

**Review by sguilhen — DISMISSED** [review]
Looks good @mhajas, I've spotted only very minor issues that don't prevent me from approving this.

**Comment by mhajas** [comment]
Thanks for the review @sguilhen I fixed these.

**Review by hmlnarik — APPROVED** [review]
Thanks @mhajas !

**Pull request merged**